File: apps/ml/ui/src/types/wallet.ts
export interface Wallet {
  id: string
  user_email: string
  balance: number
  reserved: number
  currency: string
  local_quota_seconds: number
  local_used_seconds: number
  local_quota_reset_at: string | null
  created_at: string
  updated_at: string
}

export interface WalletTransaction {
  id: string
  type: 'credit' | 'debit' | 'reserve' | 'release'
  amount: number
  balance_after: number
  reserved_after: number
  description: string
  reference: string | null
  job_id: string | null
  created_at: string
}

export interface PlatformLedgerEntry {
  id: string
  amount_usd: number
  recipient_email: string
  performed_by: string
  note: string
  wallet_tx_id: string
  created_at: string
}

export interface AdminLedgerResponse {
  items: PlatformLedgerEntry[]
  total: number
  total_spent_usd: number
  page: number
  page_size: number
}

export interface AdminUserSummary {
  email: string
  full_name: string
  org_id: string
  role: string
}

export interface LocalQuota {
  quota_seconds: number
  used_seconds: number
  remaining_seconds: number
  quota_hours: number
  used_hours: number
  remaining_hours: number
  reset_at: string | null
  exhausted: boolean
}

File: apps/ml/app/models/wallet.py
from datetime import datetime, timezone
from typing import Optional
from beanie import Document
from pydantic import Field


def utc_now():
    return datetime.now(timezone.utc)


class Wallet(Document):
    class Settings:
        name = "wallets"

    org_id: str = ""
    user_email: str
    balance: float = 0.0      # total available (not including reserved)
    reserved: float = 0.0     # held for pending GPU jobs
    currency: str = "USD"

    # Monthly local training quota
    local_quota_seconds: int = 36_000      # 10 hrs free/month (configurable via purchase)
    local_used_seconds: float = 0.0        # seconds consumed this billing cycle
    local_quota_reset_at: Optional[datetime] = None  # when usage resets (start of next month)

    created_at: datetime = Field(default_factory=utc_now)
    updated_at: datetime = Field(default_factory=utc_now)


class WalletTransaction(Document):
    class Settings:
        name = "wallet_transactions"

    org_id: str = ""
    user_email: str
    type: str   # credit | debit | reserve | release
    amount: float
    balance_after: float
    reserved_after: float
    description: str
    reference: Optional[str] = None   # Paystack reference
    job_id: Optional[str] = None
    created_at: datetime = Field(default_factory=utc_now)
